From fd34ffb031f627e470f6488175a611d1368933b1 Mon Sep 17 00:00:00 2001
From: Larry Woodman <lwoodman@redhat.com>
Date: Fri, 30 May 2014 11:22:34 -0400
Subject: [kernel] locking/mutexes: Return false if task need_resched() in mutex_can_spin_on_owner()

Message-id: <1401448958-5278-7-git-send-email-lwoodman@redhat.com>
Patchwork-id: 81233
O-Subject: [RHEL7.1 PATCH 06/10] locking/mutexes: Return false if task need_resched() in mutex_can_spin_on_owner()
Bugzilla: 1087655 1087919 1087922
RH-Acked-by: Rafael Aquini <aquini@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

commit 46af29e479cc0c1c63633007993af5292c2c3e75
 Author: Jason Low <jason.low2@hp.com>
 Date:   Tue Jan 28 11:13:12 2014 -0800

    locking/mutexes: Return false if task need_resched() in mutex_can_spin_on_owner()

    The mutex_can_spin_on_owner() function should also return false if the
    task needs to be rescheduled to avoid entering the MCS queue when it
    needs to reschedule.

    Signed-off-by: Jason Low <jason.low2@hp.com>
    Signed-off-by: Peter Zijlstra <peterz@infradead.org>
    Cc: Waiman.Long@hp.com
    Cc: torvalds@linux-foundation.org
    Cc: tglx@linutronix.de
    Cc: riel@redhat.com
    Cc: akpm@linux-foundation.org
    Cc: davidlohr@hp.com
    Cc: hpa@zytor.com
    Cc: andi@firstfloor.org
    Cc: aswin@hp.com
    Cc: scott.norton@hp.com
    Cc: chegu_vinod@hp.com
    Cc: paulmck@linux.vnet.ibm.com
    Link: http://lkml.kernel.org/r/1390936396-3962-2-git-send-email-jason.low2@hp.com
    Signed-off-by: Ingo Molnar <mingo@kernel.org>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/mutex.c b/kernel/mutex.c
index ca22543..e79515f 100644
--- a/kernel/mutex.c
+++ b/kernel/mutex.c
@@ -171,6 +171,9 @@ static inline int mutex_can_spin_on_owner(struct mutex *lock)
 {
  int retval = 1;
 
+ if (need_resched())
+  return 0;
+
  rcu_read_lock();
  if (lock->owner)
   retval = lock->owner->on_cpu;
-- 
1.7.1